dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java

```diff
@@ -311,6 +308,23 @@ boolean handlePropertyHits(BackedSession d2) {
                     }
                 }
             }
+
+            // Update the session's main cache entry per the identified changes
+            if (propsToWrite != null || propsToRemove != null) {
+                String sessionKey = createSessionKey(id, appName);
+                ArrayList<?> oldValue, newValue;
+                do {
+                    oldValue = cacheStoreService.cache.get(sessionKey);
+                    if (oldValue == null)
+                        throw new UnsupportedOperationException(); // TODO implement code path where cache entry for session is expired. Delete the property entries?
+                    SessionInfo sessionInfo = new SessionInfo(oldValue).clone();
+                    if (propsToWrite != null)
+                        sessionInfo.addSessionPropertyIds(propsToWrite);
+                    if (propsToRemove != null)
+                        sessionInfo.removeSessionPropertyIds(propsToRemove);
+                    newValue = sessionInfo.getArrayList();
+                } while (!cacheStoreService.cache.replace(sessionKey, oldValue, newValue));
```

Comment:
I'm worried about two clients updating the same cache and looping indefinitely here.  We should write a test that tries to force this situation.  Also, we should implement a max-retry count here where we blow up after something like 3 retries, in order to sniff out whether or not this situation actually occurs.